### PR TITLE
SF-1666 - The Avatar icon displayed as the first number of phone number

### PR DESCRIPTION
**Pull request by nigel-wells**
- Only applies to avatars currently served from Auth0 CDN
 - Supports Gravatar when an email address is available
 - Only allows for 2 characters - Auth0 limitation

### CODE REVIEW

**Review by RaymondLuong3 — CHANGES_REQUESTED** [review]
Nice work. I have a few comments.

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-:-NBOZcmxB68lDZbwcF_t:bjw9t03)* status: all files reviewed, 7 unresolved discussions (waiting on @nigel-wells)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOTbHD1Z-Ah2IQnPJN):*
A better commit message would describe what this commit does, rather than the description of the issue. So something like "Improve avatar icon for phone number users" or "Use name for avatar for phone number users when provided"



_Code quote:_
```
SF-1666 - The Avatar icon displayed as the first number of phone number
```
___
*[`src/SIL.XForge/Services/UserService.cs` line 159 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOUo7O0NLbbFaUGgz3:-NBOUo7O0NLbbFaUGgz4:bbvar6) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Services/UserService.cs#L159)):*
> ```Smalltalk
>                 // Only overwrite the avatar for allowed domains so as not to overwrite an avatar provided by a social connection
>                 string[] excludeDomains = { "cdn.auth0.com", "gravatar.com" };
>                 if (!excludeDomains.Any(userDoc.Data.AvatarUrl.Contains))
> ```

I think this is incomplete ;)



_Code quote:_
```Smalltalk
                if (!excludeDomains.Any(userDoc.Data.AvatarUrl.Contains))
```
___
*[`src/SIL.XForge/Services/UserService.cs` line 164 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOVJY1-4Hto4lczovH:-NBOVJY1-4Hto4lczovI:bdv0bfa) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Services/UserService.cs#L164)):*
> ```Smalltalk
>                 }
> 
>                 var initials = string.Concat(
> ```

Typically we only use var if the type is explicit. Use string[] here.



_Code quote:_
```Smalltalk
                var initials = string.Concat(
```
___
*[`src/SIL.XForge/Services/UserService.cs` line 183 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOVcjw31jNyjHTwQgE:-NBOVcjw31jNyjHTwQgF:blxl3cy) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Services/UserService.cs#L183)):*
> ```Smalltalk
>                 if (userDoc.Data.Email != null && userDoc.Data.Email != "")
>                 {
>                     var emailHash = StringUtils.ComputeMd5Hash(userDoc.Data.Email);
> ```

I had no idea this is how email gravatars were created. Nice.



_Code quote:_
```Smalltalk
            var emailHash = StringUtils.ComputeMd5Hash(userDoc.Data.Email);
```
___
*[`src/SIL.XForge/Utils/StringUtils.cs` line 22 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOVx7rAhxnlonO44JO:-NBOVx7rAhxnlonO44JP:b-yue6z5) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Utils/StringUtils.cs#L22)):*
> ```Smalltalk
>                 for (int i = 0; i < hash.Length; i++)
>                 {
>                     sb.Append(hash[i].ToString("X2"));
> ```

What does the "X2" formatting do to the string?



_Code quote:_
```Smalltalk
 sb.Append(hash[i].ToString("X2"));
```
___
*[`test/SIL.XForge.Tests/Services/UserServiceTests.cs` line 123 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOX9IY9DczJCCTXZRa:-NBOX9IY9DczJCCTXZRb:bktkyvm) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/test/SIL.XForge.Tests/Services/UserServiceTests.cs#L123)):*
> ```Smalltalk
>             IDocument<User> userDoc = await conn.FetchAsync<User>(userId);
> 
>             string[,] expectedInitials =
> ```

I am not familiar with this syntax. Does this mean it is an array of string tuples? Great!



_Code quote:_
```Smalltalk
string[,]
```
___
*[`test/SIL.XForge.Tests/Services/UserServiceTests.cs` line 138 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOYhtZ7LgZyulTH36c:-NBOYhtZ7LgZyulTH36d:b8gxxz8) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/test/SIL.XForge.Tests/Services/UserServiceTests.cs#L138)):*
> ```Smalltalk
> 
>             // Check avatar URL is updated - only happens if the existing URL is one set by Auth0
>             for (int i = 0; i < expectedInitials.GetLength(0); i++)
> ```

Did `expectedIntialis.Length` not work in this situation. Just curious.



_Code quote:_
```Smalltalk
  for (int i = 0; i < expectedInitials.GetLength(0); i++)
```

**Review by nigel-wells — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1488)* status: 8 of 9 files reviewed, 7 unresolved discussions (waiting on @RaymondLuong3)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOTbHD1Z-Ah2IQnPJN):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

A better commit message would describe what this commit does, rather than the description of the issue. So something like "Improve avatar icon for phone number users" or "Use name for avatar for phone number users when provided"
</blockquote></details>

Ah yep, that's a habit I need to get out. I've renamed it now and will try to remember in the future.
___
*[`src/SIL.XForge/Services/UserService.cs` line 159 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOUo7O0NLbbFaUGgz3:-NBQ0uen1JECLij0CyWr:b-b20is8) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Services/UserService.cs#L159)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I think this is incomplete ;)
</blockquote></details>

Whoops! It used to be that but then I realised the `AvatarUrl` could be so many different things so I reversed it. I've renamed to `allowedDomains`
___
*[`src/SIL.XForge/Services/UserService.cs` line 164 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOVJY1-4Hto4lczovH:-NBQ1TCm4PPEJC0wmFKR:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Services/UserService.cs#L164)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Typically we only use var if the type is explicit. Use string[] here.
</blockquote></details>

Done.
___
*[`src/SIL.XForge/Services/UserService.cs` line 183 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOVcjw31jNyjHTwQgE:-NBQ1WIp1DD_etrwCndn:b-v6z4pd) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Services/UserService.cs#L183)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I had no idea this is how email gravatars were created. Nice.
</blockquote></details>

Yeah nice learning for me as well!
___
*[`src/SIL.XForge/Utils/StringUtils.cs` line 22 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOVx7rAhxnlonO44JO:-NBQ3yjp4NeIaJvu2FW4:bxee6jz) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Utils/StringUtils.cs#L22)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

What does the "X2" formatting do to the string?
</blockquote></details>

I'm not too clued up when it comes to the logic around hashes. I believe it is just taking the byte generated and converting it to a specific format of string that is needed for the MD5 encryption. I'm sure there is a read up somewhere. I found this code on a number of sites and they all did exactly the same thing when it came to generated an MD5 hash encryption.
___
*[`test/SIL.XForge.Tests/Services/UserServiceTests.cs` line 123 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOX9IY9DczJCCTXZRa:-NBQ2L6r8hBXal0ylCHk:b-ojhwsc) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/test/SIL.XForge.Tests/Services/UserServiceTests.cs#L123)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I am not familiar with this syntax. Does this mean it is an array of string tuples? Great!
</blockquote></details>

Yep that's how I understand the multidimensional array. You can read more here: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/multidimensional-arrays
___
*[`test/SIL.XForge.Tests/Services/UserServiceTests.cs` line 138 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOYhtZ7LgZyulTH36c:-NBQ2RfU8QR6K2jzn6iI:b-ter36w) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/test/SIL.XForge.Tests/Services/UserServiceTests.cs#L138)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Did `expectedIntialis.Length` not work in this situation. Just curious.
</blockquote></details>

This is because of the multidimension array I used. `Length` is the total number of items throughout the entire array which in my case would be 16 (8 rows * 2 dimensions). `GetLength(0)` counts the number of items in the `0` dimension of the array which is 8.

**Review by nigel-wells — COMMENTED** [review]
This is building again now if you want to finish your review.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-:-NBor5kW7Z8EoO4AKfg1:b-vdp3a)* status: 8 of 9 files reviewed, 7 unresolved discussions (waiting on @RaymondLuong3)

**Review by RaymondLuong3 — CHANGES_REQUESTED** [review]
Thanks for your work. This looks to be working for me.

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-:-NBryum0Bty1bN8g9CiM:bcmrho1)* status: all files reviewed, 1 unresolved discussion (waiting on @nigel-wells)

___
*[`src/SIL.XForge/Services/UserService.cs` line 159 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBrtQoL9YxNVUT3dUyj:-NBrtQoL9YxNVUT3dUyk:b-txujrr) ([raw file](https://github.com/sillsdev/web-xforge/blob/4551c87d429a932faad5150e9616167ac687bc03/src/SIL.XForge/Services/UserService.cs#L159)):*
> ```Smalltalk
>                 // Only overwrite the avatar for allowed domains so as not to overwrite an avatar provided by a social connection
>                 string[] allowedDomains = { "cdn.auth0.com", "gravatar.com" };
>                 if (!allowedDomains.Any(userDoc.Data.AvatarUrl.Contains))
> ```

Do you mean to say `if (!allowedDomains.Any(a => userDoc.Data.AvatarUrl.Contains(a))`?



_Code quote:_
```Smalltalk
if (!allowedDomains.Any(userDoc.Data.AvatarUrl.Contains))
```
___
*[`src/SIL.XForge/Utils/StringUtils.cs` line 22 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOVx7rAhxnlonO44JO:-NBruvHE1t4aVSvSnpkO:by3scbc) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/src/SIL.XForge/Utils/StringUtils.cs#L22)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

I'm not too clued up when it comes to the logic around hashes. I believe it is just taking the byte generated and converting it to a specific format of string that is needed for the MD5 encryption. I'm sure there is a read up somewhere. I found this code on a number of sites and they all did exactly the same thing when it came to generated an MD5 hash encryption.
</blockquote></details>

Ok, that is good to know this is how the examples said to use.
___
*[`test/SIL.XForge.Tests/Services/UserServiceTests.cs` line 123 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOX9IY9DczJCCTXZRa:-NBrv3nZBleZjKAwJMNa:b-83sjo4) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/test/SIL.XForge.Tests/Services/UserServiceTests.cs#L123)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

Yep that's how I understand the multidimensional array. You can read more here: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/multidimensional-arrays
</blockquote></details>

Nice.
___
*[`test/SIL.XForge.Tests/Services/UserServiceTests.cs` line 138 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBOYhtZ7LgZyulTH36c:-NBrvA-gEJZDhEZxB_VC:b-mbdy26) ([raw file](https://github.com/sillsdev/web-xforge/blob/af56abf423f47b85be9242a2aa377779265eb145/test/SIL.XForge.Tests/Services/UserServiceTests.cs#L138)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

This is because of the multidimension array I used. `Length` is the total number of items throughout the entire array which in my case would be 16 (8 rows * 2 dimensions). `GetLength(0)` counts the number of items in the `0` dimension of the array which is 8.
</blockquote></details>

Nice, I'm re-learning something I haven't done in a while.

**Review by nigel-wells — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1488)* status: 8 of 9 files reviewed, 1 unresolved discussion (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge/Services/UserService.cs` line 159 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBrtQoL9YxNVUT3dUyj:-NCwVExD5wOtK5Kt0Mwx:byomkb) ([raw file](https://github.com/sillsdev/web-xforge/blob/4551c87d429a932faad5150e9616167ac687bc03/src/SIL.XForge/Services/UserService.cs#L159)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Do you mean to say `if (!allowedDomains.Any(a => userDoc.Data.AvatarUrl.Contains(a))`?
</blockquote></details>

What I've written is the shorthand for what you're suggesting and appears to also be a recommended answer on stackoverflow and other places - I can change if you feel it is still required.

**Review by Nateowami — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1488)* status: 8 of 9 files reviewed, 1 unresolved discussion (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge/Services/UserService.cs` line 159 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBrtQoL9YxNVUT3dUyj:-NCwqFcXEliYAxP4_E2X:bf5k47i) ([raw file](https://github.com/sillsdev/web-xforge/blob/4551c87d429a932faad5150e9616167ac687bc03/src/SIL.XForge/Services/UserService.cs#L159)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

What I've written is the shorthand for what you're suggesting and appears to also be a recommended answer on stackoverflow and other places - I can change if you feel it is still required.
</blockquote></details>

I think this syntax is common with functional programming, but in JavaScript we tend to avoid it, since it messes with the value of `this`.

In other words, if this feels wrong, I think it's because JavaScript has taught us to avoid this, due to it's unusual behavior. If this was JavaScript, I'd say definitely don't do this, but I think it's normal in C#.

**Review by RaymondLuong3 — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1488)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nigel-wells)

___
*[`src/SIL.XForge/Services/UserService.cs` line 159 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1488#-NBrtQoL9YxNVUT3dUyj:-NCzfHVS60YWjmkRtVOW:b7wxdrl) ([raw file](https://github.com/sillsdev/web-xforge/blob/4551c87d429a932faad5150e9616167ac687bc03/src/SIL.XForge/Services/UserService.cs#L159)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

I think this syntax is common with functional programming, but in JavaScript we tend to avoid it, since it messes with the value of `this`.

In other words, if this feels wrong, I think it's because JavaScript has taught us to avoid this, due to it's unusual behavior. If this was JavaScript, I'd say definitely don't do this, but I think it's normal in C#.
</blockquote></details>

Ah ok. It has been a while since I've seen that syntax, passing in a named method rather than an anonymous function. Good work
